src/descriptor/tr.rs

```diff
@@ -0,0 +1,698 @@
+    /// If spend data is already computed (i.e it is not None), this does not recompute it
+    /// TaprootSpendInfo is only required for spending via the script paths.
+    pub fn spend_info<C: secp256k1::Verification>(
+        &mut self,
+        secp: &secp256k1::Secp256k1<C>,
+    ) -> &TaprootSpendInfo
+    where
+        Pk: ToPublicKey,
+    {
+        // Already computed cache
+        if let Some(ref info) = self.spend_info {
+            return info;
+        } else {
+            // Key spend path with no merkle root
+            if self.tree.is_none() {
+                let data = TaprootSpendInfo::new_key_spend(
+                    secp,
+                    self.internal_key.to_x_only_pubkey(),
+                    None,
+                );
+                self.spend_info = Some(data);
+                return self.spend_info.as_ref().unwrap();
+            }
+            let mut builder = TaprootBuilder::new();
+            for (depth, ms) in self.iter_scripts() {
+                let script = ms.encode();
+                builder = builder
+                    .add_leaf(depth, script)
+                    .expect("Computing spend data on a valid Tree should always succeed");
+            }
+            // Assert builder cannot error here
+            let data = match builder.finalize(secp, self.internal_key.to_x_only_pubkey()) {
+                Ok(data) => data,
+                Err(e) => match e {
+                    TaprootBuilderError::InvalidMerkleTreeDepth(_) => {
+                        unreachable!("Depth checked in struct construction")
+                    }
+                    TaprootBuilderError::NodeNotInDfsOrder => {
+                        unreachable!("Insertion is called in DFS order")
+                    }
+                    TaprootBuilderError::OverCompleteTree => {
+                        unreachable!("Taptree is a well formed tree")
+                    }
+                    TaprootBuilderError::InvalidInternalKey(_) => {
+                        unreachable!("Internal key checked for validity")
+                    }
+                    TaprootBuilderError::IncompleteTree => {
+                        unreachable!("Taptree is a well formed tree")
+                    }
+                    TaprootBuilderError::EmptyTree => {
+                        unreachable!("Taptree is a well formed tree with atleast 1 element")
+                    }
+                },
+            };
+            // Add the data
+            self.spend_info = Some(data);
+            return &self.spend_info.as_ref().unwrap();
+        }
+    }
+}
+
+impl<Pk: MiniscriptKey + ToPublicKey> Tr<Pk> {
+    /// Obtain the corresponding script pubkey for this descriptor
+    /// Same as[`DescriptorTrait::script_pubkey`] for this descriptor
+    pub fn spk(&self) -> Result<Script, Error> {
+        let spend_info = self
+            .spend_info
+            .as_ref()
+            .ok_or(Error::TaprootSpendInfoUnavialable)?;
```

Comment:
This error case is the thing that forces as to make `script_pubkey()` return a `Result` right? Wouldn't it be preferable to make caching `spend_info` optional. So make a public API for caching it but if it hasn't been cached just compute it on the fly. For me, the risk of possibly doing some extra (usually minor) computation time seems better than giving me a `Result` that I may arrogantly `.unwrap()`.

Reply:
This is not possible because we require a `secp` parameter to do the tweaking. Creating a new context inside a method is expensive and causes severe performance issues. This would go away once we get static verification tables in rust-secp, but still a while to go.

I am also unhappy with the current PR :( . Some rough thoughts, 

- I am considering more changes in the type system to distinguish between `Derived` and `UnDerived` descriptors. We would then implement `DescriptorTriat` only for `Derived` descriptors. Or perhaps having two traits.
- I also don't like that `address()` fails for raw descriptors which are not that common. Maybe `DescriptorTrait` methods are not that useful. 

I will make another candidate PR by tomorrow with the goal of avoiding unwraps() for common case like `script_pubkey` and `address` etc